File: Python/102_BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """

        que = []

        if root == None:
            return []

        que.append(root)
        values = []
        res = []

        while len(que) > 0:

            for i in range(len(que)):
                node = que.pop(0)
                if node != None:
                    values.append(node.val)
    
                if node != None:
                    if node.left != None:
                        que.append(node.left)
                    if node.right != None:
                        que.append(node.right)

            res.append(values)
            values = []

        return res

# Algorithm: BSF
# store nodes in each depth
# exceptional case: root is none        

File: C++/102_BinaryTreeLevelOrderTraversal.cpp
using namespace std;
#include <vector>
#include <queue>

 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
 
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> answers;
        vector<int> ans;
        queue<TreeNode*> q;

        if (!root) return {};
        q.push(root);

        while(!q.empty()){
            int count = q.size();

            for(size_t i=0;i<count;i++) {
                TreeNode* cur = q.front();
                q.pop();
                if (cur) {
                    ans.push_back(cur->val);
                    if (cur->left) q.push(cur->left);
                    if (cur->right) q.push(cur->right);
                }
            }

            answers.push_back(ans);
            while(!ans.empty()) ans.pop_back();

        }

        return answers;        
    }
};